independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java

```diff
@@ -1089,6 +1090,41 @@ private void implementCreateForSyntheticBean(ClassCreator beanCreator, BeanInfo
                 .setModifiers(ACC_PRIVATE);
         bean.getCreatorConsumer().accept(createSynthetic);
 
+        Consumer<MethodCreator> checkActive = bean.getCheckActiveConsumer();
+        if (checkActive != null) {
+            MethodCreator checkActiveMethod = beanCreator.getMethodCreator("checkActive", void.class)
+                    .setModifiers(ACC_PRIVATE);
+            checkActive.accept(checkActiveMethod);
+
+            List<InjectionPointInfo> matchingIPs = new ArrayList<>();
+            for (InjectionPointInfo injectionPoint : bean.getDeployment().getInjectionPoints()) {
+                if (bean.equals(injectionPoint.getResolvedBean())) {
+                    matchingIPs.add(injectionPoint);
+                }
+            }
+
+            TryBlock tryBlock = doCreate.tryBlock();
+            tryBlock.invokeSpecialMethod(checkActiveMethod.getMethodDescriptor(), tryBlock.getThis());
+            CatchBlockCreator catchBlock = tryBlock.addCatch(Exception.class);
```

Comment:
I talked about this with @mkouba last week and we eventually concluded that a `Runnable` throwing an exception is a bad API. The proper one should be something like `Supplier<ActiveResult>`, where

```java
public final class ActiveResult {
    private final boolean active;
    private final String inactiveReason;

    public static ActiveResult active() {
        return new ActiveResult(true, null);
    }

    public static ActiveResult inactive(String reason) {
        return new ActiveResult(false, Objects.requireNonNull(reason));
    }

    private ActiveResult(boolean active, String inactiveReason) {
        this.active = active;
        this.inactiveReason = inactiveReason;
    }

    public boolean isActive() {
        return active;
    }

    public String inactiveReason() {
        return inactiveReason;
    }
}
```

If the supplier throws, we won't catch that and everything will blow up, which is probably better than what we have here.

Reply:
Can we add a cause into this? It's possible one bean would be inactive because another is inactive, e.g. Liquibase only makes sense if the relevant datasource is active, and in that case you'd want to forward the `InactiveBeanException` thrown by accessing the datasource.

Reply:
That sounds reasonable. I'm just wondering if the `cause` should be arbitrary `Exception`, or just `InactiveBeanException`...

Reply:
Either way works I suppose, as long as the cause's stack trace is displayed (so that users can see the cause of the cause).

Reply:
Hey @Ladicek,

So I've got some more input here... With the current code I ended up doing this:

```java
    public Supplier<ActiveResult> flywayActiveSupplier(String dataSourceName) {
        return new Supplier<ActiveResult>() {
            @Override
            public ActiveResult get() {
                FlywayDataSourceRuntimeConfig dataSourceConfig = config.getValue()
                        .getConfigForDataSourceName(dataSourceName);

                // Flyway can only be active when the datasource itself is active.
                try {
                    ClientProxy.unwrap(AgroalDataSourceUtil.dataSourceInstance(dataSourceName).get());
                } catch (InactiveBeanException e) {
                    if (dataSourceConfig.active.isPresent() && dataSourceConfig.active.get()) {
                        String propertyKey = FlywayContainerUtil.getFlywayPropertyKey(dataSourceName, "active");
                        throw new ConfigurationException(String.format(Locale.ROOT,
                                "Flyway for datasource '%1$s' was activated explicitly through configuration property '%2$s',"
                                        + " but datasource '%1$s' was deactivated."
                                        + " Either deactivate Flyway, or activate the datasource.",
                                dataSourceName, propertyKey),
                                Set.of(propertyKey));
                    }
                    return ActiveResult.inactive(String.format(Locale.ROOT,
                            "Flyway for datasource '%s' was deactivated automatically because this datasource was deactivated."
                                    + e.getMessage(),
                            dataSourceName));
                }

                // Even if the datasource is active, Flyway can still be deactivated explicitly
                if (dataSourceConfig.active.isPresent() && !dataSourceConfig.active.get()) {
                    return ActiveResult.inactive(String.format(Locale.ROOT,
                            "Flyway for datasource '%s' was deactivated through configuration properties."
                                    + " To keep Flyway deactivated and avoid this error:"
                                    + " inject Flyway-dependent CDI beans using Instance<Sometype>' instead of 'Sometype',"
                                    + " and make sure code only accesses these beans if Flyway actually is active."
                                    + " To activate Flyway: set configuration property '%s' to 'true'.",
                            dataSourceName, FlywayContainerUtil.getFlywayPropertyKey(dataSourceName, "active")));
                }

                return ActiveResult.active();
            }
        };
    }
```

As you can see, I need to detect that the datasource is inactive, because in that case Flyway must be deactivated, and to do that I must to try to retrieve the datasource, because that's the only way I can forward to the user the reason for the datasource (and Flyway) being deactivated.

That's the best I can do right now, but it's still not ideal, because the inactive reason for Flyway would include the full message of the `InactiveBeanException` for the datasource... including among other the list of beans that get injected with a `DataSource`, which is really not relevant when the problem is the injection point of a `Flyway` bean.

There's also the problem that the "inactive reason" for a datasource needs to be actionable, so in that reason I had to include advice such as "make sure you inject the datasource as `Instance<DataSource>` instead of just `DataSource`"... which really isn't relevant in the "inactive reason" of a `Flyway` bean -- errors involving that bean would not involve the injection of a `DataSource`.

To solve the above, I would suggest:

1. Changing `ActiveResult.inactive(String reason)` to `ActiveResult.inactive(Exception cause)`.
2. Preserving the existing way to build the `InactiveBeanException` message, just replacing the `.append(reason)` with `.append(cause.getMessage())`.
3. Adding more hints regarding how to avoid the `InactiveBeanException` in the `InactiveBeanException` message. See my other comment.
4. Always assigning the cause to `InactiveBeanException`.

With 1, 2 and 3, we make sure that users getting an `InactiveBeanException` will get all the information they need to get out of this situation -- with an actionable message, and a detailed stacktrace should they need it.
With 3, we make sure that extensions declaring inactive beans don't need to include information in their message that is irrelevant to consumers of "transitively deactivated" beans (e.g. Flyway)
With 4, we make sure that extensions (e.g. Flyway) trying to propagate the reason that another bean (e.g. DataSource) is inactive will have access to the full message, without any irrelevant information: they just need to call `inactiveBeanException.getCause().getMessage()`.

Reply:
It really feels like an inactive `ActiveReason` should be able to hold a cause, which would be another inactive `ActiveReason` (potentially recursively). We could then print the whole cause chain in the `InactiveBeanException` message.

To be able to access the causing `ActiveReason`, I ended up changing `InjectableBean.isActive()` from returning `boolean` to returning `ActiveResult`.

You should be able to do something like:

```java
ActiveResult active = AgroalDataSourceUtil.dataSourceInstance(dataSourceName).getHandle().getBean().isActive();
```

And then pass `active` to any `ActiveResult.inactive()` calls you make. The `InactiveBeanException` message then looks like:

```
Bean is not active: ...
Reason: top-level reason
Cause: first-level cause reason
Cause: second-level cause reason
...
```

WDYT? Does that make sense, or is exception chaining really better?

Reply:
Works for me, thanks! I'll try that.

You may want to change `isActive` to `getActive`, since the convention is usually to return a `boolean` from a `is*` method, but I don't _personally_ mind.

I could see exception chaining being better to display a proper stack trace, especially if a bean's inactivity is caused by a user exception for some reason. But right now I don't have such case, so `ActiveResult` chaining sounds good.

Reply:
FWIW, this works perfectly in my WIP PR, though you may want to have a look at my comments:

> You may want to change isActive to getActive, since the convention is usually to return a boolean from a is* method, but I don't personally mind.
> 
> I could see exception chaining being better to display a proper stack trace, especially if a bean's inactivity is caused by a user exception for some reason. But right now I don't have such case, so ActiveResult chaining sounds good.

Up to you!

Reply:
I've renamed the `ActiveResult`-returning method `isActive()` to `checkActive()`. I've also added a `boolean`-returning method `isActive()`, which does just `return checkActive().value();`.

Which reminds me that I've also renamed `ActiveResult.result()` to `value()`, just for better readability.

I've left out exception chaining. I guess that's something we might add in the future, if the need comes.